renku/core/project/project_properties.py

```diff
@@ -17,68 +17,364 @@
+
+    @datadir.setter
+    def datadir(self, value: str):
+        """Set the current datadir."""
+        self._top.datadir = value
+
+    @property
+    def dataset_images_path(self) -> Path:
+        """Return a ``Path`` instance of Renku dataset metadata folder."""
+        return self.path / RENKU_HOME / DATASET_IMAGES
 
     @property
-    def path(self):
+    def docker_path(self):
+        """Path to the Dockerfile."""
+        return self.path / DOCKERFILE
+
+    @property
+    def empty(self) -> bool:
+        """Return True if there is no pushed context."""
+        return not bool(self._context_stack)
+
+    @property
+    def global_config_dir(self) -> str:
+        """Return user's config directory."""
+        import click
+
+        return click.get_app_dir(APP_NAME, force_posix=True)
+
+    @property
+    def global_config_path(self) -> Path:
+        """Renku global (user's) config path."""
+        config = Path(self.global_config_dir)
+        if not config.exists():
+            config.mkdir(parents=True)
+
+        return config / CONFIG_NAME
+
+    @property
+    def global_config_read_lock(self):
+        """Create a user-level config read lock."""
+        from renku.core.util.contexts import Lock
+
+        return Lock(self.global_config_path)
+
+    @property
+    def global_config_write_lock(self):
+        """Create a user-level config write lock."""
+        from renku.core.util.contexts import Lock
+
+        return Lock(self.global_config_path, mode="exclusive")
+
+    @property
+    def latest_agent(self) -> Optional[str]:
+        """Returns latest agent version used in the repository."""
+        try:
+            return self.project.agent_version
+        except ValueError as e:
+            if "Project name not set" in str(e):
+                return None
+            raise
+
+    @property
+    def local_config_path(self) -> Path:
+        """Renku local (project) config path."""
+        return self.metadata_path / CONFIG_NAME
+
+    @property
+    def lock(self):
+        """Create a Renku config lock."""
+        import filelock
+
+        return filelock.FileLock(str(self.metadata_path.with_suffix(LOCK_SUFFIX)), timeout=0)
+
+    @property
+    def metadata_path(self) -> Path:
+        """Current project's metadata (RENKU_HOME) path."""
+        return self.path / RENKU_HOME
+
+    @property
+    def path(self) -> Path:
         """Current project path."""
-        return self._path_stack[-1]
+        return self._top.path
+
+    @property
+    def pointers_path(self) -> Path:
+        """Return a ``Path`` instance of Renku pointer files folder."""
+        path = self.path / RENKU_HOME / POINTERS
+        path.mkdir(exist_ok=True)
+        return path
+
+    @property
+    def project(self) -> "Project":
+        """Return the Project instance."""
+        if not self._top.project:
+            try:
+                self._top.project = self.database["project"]  # type: ignore
```

Comment:
`IProjectGateway.get_project()` uses `project_context.database` to return the project:
```
        try:
            return project_context.database["project"]  # type: ignore
        except KeyError as e:
            raise ValueError() from e
```

We also decided to remove `IProjectGateway.get_project()` and rely on the `project_context` to get the project (which I'm doing it in my current PR).

Reply:
We should do it the other way around. We should not access the database directly to get these things, except for in the gateway. otherwise we have a strong coupling to the database in places that don't need it. There should be an issue with project_context.project -> ProjecGateway.get_project -> project_context.database to get it, it's not really circular and project_context is initialized at this point.

Reply:
I see your point. I've change it as you've suggested.